cpp/src/arrow/device.cc

```diff
@@ -100,6 +100,20 @@ Result<std::unique_ptr<Buffer>> MemoryManager::CopyNonOwned(
                                 " to ", to->device()->ToString(), " not supported");
 }
 
+Status MemoryManager::CopyBufferSliceFrom(const std::shared_ptr<Buffer>& buf, void* to,
+                                          const int64_t offset, const int64_t length) {
+  auto maybe_buffer = CopyBuffer(buf, default_cpu_memory_manager());
+
+  if (COPY_BUFFER_SUCCESS(maybe_buffer)) {
+    memcpy(to, (*maybe_buffer)->data() + offset, length);
+    return Status::OK();
+  }
+
+  return Status::NotImplemented("Copying buffer slice from ",
+                                buf->memory_manager()->device()->ToString(),
+                                " to memory not supported");
+}
+
```

Comment:
This class is complicated, so I had to try implementing this to understand the code.

One of the things I've learned: the `-From` suffix means your function takes `from` `MemoryManager` parameter that matches `buf->memory_manager()`. That doesn't apply here.

I also realized that the `static` function is enough, no need to have the `virtual` version — the fact that we are always copying to CPU memory means we don't have to virtual-dispatch another call and we only have to virtual-dispatch to `buf->memory_manager()`.

So here is what I drafted (I didn't test it):

```cpp
Status MemoryManager::CopyBufferSlice(const std::shared_ptr<Buffer>& buf, int64_t offset,
                                      int64_t length, uint8_t* out_data) {
  if (ARROW_PREDICT_TRUE(buf->is_cpu())) {
    memcpy(out_data, buf->data() + offset, static_cast<size_t>(length));
    return Status::OK();
  }

  auto& from = buf->memory_manager();
  auto cpu_mm = default_cpu_memory_manager();
  // Try a view first
  auto maybe_buffer_result = from->ViewBufferTo(buf, cpu_mm);
  if (!COPY_BUFFER_SUCCESS(maybe_buffer_result)) {
    // View failed, try a copy instead
    maybe_buffer_result = from->CopyBufferTo(buf, cpu_mm);
  }
  ARROW_ASSIGN_OR_RAISE(auto maybe_buffer, std::move(maybe_buffer_result));
  if (maybe_buffer != nullptr) {
    memcpy(out_data, maybe_buffer->data() + offset, static_cast<size_t>(length));
    return Status::OK();
  }

  return Status::NotImplemented("Copying buffer slice from ", from->device()->ToString(),
                                " to CPU not supported");
}
```

Please define and declare it after `ViewBuffer` to not split that related group of functions.